Various fixes

gfontenot opened the pull request:
This includes a few things. Some minor (formatting, SwiftCheck update), some
larger (Extension API across all targets, updates for newest version of Swift
3)

Don't know how I feel about these precedence groups, tbh. I don't think I feel
super great about them. Our tests are also _completely_ broken, since it's
apparently impossible to resolve the differences between how we're choosing to
define these operators and how SwiftCheck defines them. No idea what to do
about this.

Fixes #59 
Fixes #78

gfontenot committed:
Require Extension API for all targets

We were missing this for the tvOS target.

gfontenot committed:
Fix spacing in package file

gfontenot committed:
Use @escaping properly

@noescape is now the default, so we need to be using @escaping to indicate
function params that might escape. This change was led by the compiler, which
is nice.

gfontenot committed:
Update SwiftCheck to newest version

gfontenot committed:
Use precedence groups

These remove the need for the explanitory comments about what the specific
precedence means. Unfortunately, there's no `equalTo` operator for these, so
we can't actually define these as accurately as I'd like. There's also no
ability to define multiple `higherThan` or `lowerThan` precedence groups,
which means we can't ensure that our `AlternativePrecedence` is positioned
correctly above our monadic precedences.

All in all, I'm not sure how I feel about this. I'm less confident in the
positioning of these groups now and I'm unclear on how changes will affect
consuming code.

hyperspacemark commented on Runes.xcodeproj/project.pbxproj:
```diff
@@ -713,6 +726,7 @@
 		51DE8A2B1BAB36E600124320 /* Release */ = {
 			isa = XCBuildConfiguration;
 			buildSettings = {
+				APPLICATION_EXTENSION_API_ONLY = NO;
```
Why revert this? We should be able to use Runes in extensions without warnings as Runes doesn't use any extension unsafe API.

gfontenot replied:
This is for the test bundle. I'm actually setting this to `YES` at the project level so that _all_ of the frameworks have it set properly, but the tests need to have this set to `NO`. It's a confusing diff.

gfontenot commented:
Seriously tempted to just kill the tests here and have CI simply build the framework for the various targets.

hyperspacemark commented on Sources/Runes.swift:
```diff
@@ -1,93 +1,75 @@
-    // Lower precedence than `<^>`, `<*>`, `*>`, `<*`
-    precedence 120
-}
+infix operator <|>: AlternativePrecedence
```
`AlternativePrecedence` reads as extremely vague to me. Alternative to what? Any reason to not go with, like, `LeftAssociativeBinaryPrecedence`?

gfontenot replied:
Yeah, fair point. This operator is from the `Alternative` typeclass in Haskell, which is why I named it as such. I'm totally open to dropping that baggage though.

gfontenot commented on Sources/Runes.swift:
```diff
@@ -1,93 +1,75 @@
+precedencegroup AlternativePrecedence {
+    associativity: left
+    lowerThan: ComparisonPrecedence
```
This should be `higherThan` (ComparisonPrecedence was 130, this was 140)

gfontenot commented on Sources/Runes.swift:
```diff
@@ -1,93 +1,75 @@
-    // Same precedence as the equality operator (`==`)
-    precedence 130
-}
+infix operator <^>: ComparisonPrecedence
```
Playing with this in Argo, this actually doesn't work. `ComparisonPrecedence` doesn't have an associativity so you can't actually chain these together. Fun stuff.

gfontenot committed:
Use custom precedence groups for Applicative

Using ComparisonPrecedence here didn't actually work the way we wanted.
Apparently, ComparisonPrecedence doesn't actually have an associativity
defined, so chaining these together doesn't actually work. To fix this we can
define our own precedence groups (inspired by those defined in SwiftCheck) and
set up a precedence tree.

Unfortunately since our tests are broken, it only became apparent when we
integrated this lib into Argo.

gfontenot commented on Sources/Runes.swift:
```diff
@@ -1,93 +1,85 @@
+
+precedencegroup AlternativePrecedence {
+    associativity: left
+    higherThan: ApplicativeSequencePrecedence
```
Oops I lied, this _should_ be lower.

gfontenot committed:
Reorder precedence groups

We need to explicitly put the monadic precedences at the bottom of the stack.
I'm unclear on how this will affect things.

Also fixing the precedence for Alternative to match the precedence set in
Haskell.

jshier commented:
@gfontenot I meant to post it earlier, but I ported Runes to beta 6 for my own projects, but didn't open a PR since I couldn't run the tests. Here's my [`Runes.swift` file](https://github.com/jshier/Runes/blob/master/Sources/Runes.swift). As you can see I built two custom precedence groups and used system groups for the others. I'm able to use this version just fine in my projects, replacing the earlier operator definitions.

gfontenot commented:
@jshier Ah interesting. I started running into problems when trying to use this with Argo since I needed to define the precedence of `<|` based on the applicative operators. Did you run into that?

I _think_ this is hooked up the way we'd expect it to be hooked up now. Argo seems happy, at least.

jshier commented:
@gfontenot I didn't, though I only ported the operators with the intention of matching the previous numeric levels.

gfontenot committed:
Add spaces after operator definitions

gfontenot committed:
Fix precedence for composition operator

I'm not too worried about getting this exactly right since we're not exporting
it publicly. I just want to make sure it's got the correct precedence for the
tests.

gfontenot committed:
Update SwiftCheck dependency

They removed their custom operators! Our tests build again (and should be much
less fragile moving forward)!

gfontenot committed:
Define precedence groups in relation to the stdlib

Not sure if this will help/hurt, but I can't help but feel like defining these
in terms of the stdlib instead of in their own tree will help with ambiguous
precedences.

gfontenot commented:
I think this is probably ready to roll, but I'd love a sanity check from other interested parties. ping @thoughtbot/ios

gfontenot committed:
Reduce risk of conflict for precedence groups

Turns out these need to be unique! Fun stuff. To reduce the risk of conflict
here lets go ahead and prefix our precedence group names with our package
name.

This sucks.

tonyd256 commented:
👍
